controlplane/kubeadm/controllers/controller_test.go

```diff
@@ -599,6 +603,59 @@ func TestKubeadmControlPlaneReconciler_adoption(t *testing.T) {
 	})
 }
 
+func TestKubeadmControlPlaneReconciler_SetInfraRefAnnotationToMachines(t *testing.T) {
+	g := NewWithT(t)
+
+	cluster, kcp, tmpl := createClusterWithControlPlane()
+	cluster.Spec.ControlPlaneEndpoint.Host = "bar"
+	// This will trigger scale up
+	kcp.Spec.Replicas = pointer.Int32Ptr(int32(4))
+
+	fmc := &fakeManagementCluster{
+		Machines:            internal.FilterableMachineCollection{},
+		ControlPlaneHealthy: true,
+		EtcdHealthy:         true,
+	}
+	objs := []runtime.Object{cluster.DeepCopy(), kcp.DeepCopy(), tmpl.DeepCopy()}
+	for i := 0; i < 3; i++ {
+		m, _ := createMachineNodePair(fmt.Sprintf("test-%d", i), cluster, kcp, true)
+		if i == 0 {
+			m.SetAnnotations(map[string]string{controlplanev1.MachineClonedTemplateAnnotation: "already_set"})
+		}
+		objs = append(objs, m)
+		fmc.Machines.Insert(m)
+	}
+
+	fakeClient := newFakeClient(g, objs...)
+	fmc.Reader = fakeClient
+
+	log.SetLogger(klogr.New())
+	r := &KubeadmControlPlaneReconciler{
+		Client:                    fakeClient,
+		Log:                       log.Log,
+		scheme:                    scheme.Scheme,
+		managementCluster:         fmc,
+		managementClusterUncached: fmc,
+	}
+
+	_, err := r.reconcile(context.Background(), cluster, kcp)
+	g.Expect(err).To(BeNil())
+
+	machineList := &clusterv1.MachineList{}
+	g.Expect(fakeClient.List(context.Background(), machineList, client.InNamespace(cluster.Namespace))).To(Succeed())
+	g.Expect(machineList.Items).NotTo(BeEmpty())
+	g.Expect(machineList.Items).To(HaveLen(4))
+	for _, machine := range machineList.Items {
+		g.Expect(machine.OwnerReferences).To(HaveLen(1))
+		g.Expect(machine.OwnerReferences).To(ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
+		g.Expect(machine.Labels).To(Equal(internal.ControlPlaneLabelsForClusterWithHash(cluster.Name, hash.Compute(&kcp.Spec))))
+		// Newly created machine and 3 existing machines should have the infraref annotation.
+		_, exist := machine.GetAnnotations()[controlplanev1.MachineClonedTemplateAnnotation]
+		g.Expect(exist).To(BeTrue())
```

Comment:
This pattern makes it a little hard to see what's happening when the assertion fails (the message is something to the effect of "expected false to be true"). I prefer the `g.Expect(machine.GetAnnotations()).To(HaveKey(controlplanev1.MachineClonedTemplateAnnotation))` pattern because it gives better feedback by printing out the whole map when the key is missing.

Reply:
Yes, much better. Changed.